rasa/core/lock_store.py

```diff
@@ -0,0 +1,227 @@
+
+        Perform `attempts` with a wait of `wait` seconds between them before
+        raising a `LockError`.
+        """
+
+        ticket = self.issue_ticket(conversation_id)
+
+        try:
+            # have to use async_generator.yield_() for py 3.5 compatibility
+            await yield_(
+                await self._acquire_lock(conversation_id, ticket, attempts, wait)
+            )
+        finally:
+            self.cleanup(conversation_id, ticket)
+
+    async def _acquire_lock(
+        self, conversation_id: Text, ticket: int, attempts: int, wait: Union[int, float]
+    ) -> TicketLock:
+
+        while attempts > 0:
+            # fetch lock in every iteration because lock might no longer exist
+            lock = self.get_lock(conversation_id)
+
+            # exit loop if lock does not exist anymore (expired)
+            if not lock:
+                break
+
+            # acquire lock if it isn't locked
+            if not lock.is_locked(ticket):
+                return lock
+
+            # sleep and update lock
+            await asyncio.sleep(wait)
+            self.update_lock(conversation_id)
+
+            attempts -= 1
+
+        raise LockError(
+            "Could not acquire lock for conversation_id '{}'."
+            "".format(conversation_id)
+        )
+
+    def update_lock(self, conversation_id: Text) -> None:
+        """Fetch lock for `conversation_id`, remove expired tickets and save lock."""
+
+        lock = self.get_lock(conversation_id)
+        if lock:
+            lock.remove_expired_tickets()
+            self.save_lock(lock)
+
+    def get_or_create_lock(self, conversation_id: Text) -> TicketLock:
+        """Fetch existing lock for `conversation_id` or create a new one if
+        it doesn't exist."""
+
+        existing_lock = self.get_lock(conversation_id)
+
+        if existing_lock:
+            return existing_lock
+
+        return self.create_lock(conversation_id)
+
+    def is_someone_waiting(self, conversation_id: Text) -> bool:
+        """Return whether someone is waiting for lock associated with
+        `conversation_id`."""
+
+        lock = self.get_lock(conversation_id)
+        if lock:
+            return lock.is_someone_waiting()
+
+        return False
+
+    def finish_serving(self, conversation_id: Text, ticket_number: int) -> None:
+        """Finish serving ticket with `ticket_number` for `conversation_id`.
+
+        Removes ticket from lock and saves lock.
+        """
+
+        lock = self.get_lock(conversation_id)
+        if lock:
+            lock.remove_ticket_for_ticket_number(ticket_number)
+            self.save_lock(lock)
+
+    def cleanup(self, conversation_id: Text, ticket_number: int) -> None:
+        """Remove lock for `conversation_id` if no one is waiting."""
+        self.finish_serving(conversation_id, ticket_number)
+        if not self.is_someone_waiting(conversation_id):
+            self.delete_lock(conversation_id)
```

Comment:
Could a race condition happen here?

Reply:
I don't see how - `is_someone_waiting` fetches the lock fresh from memory / storage

Reply:
Ah, and the "issue ticket - i am waiting' happens way before. Ok, agree should be fine, sorry! 👍